plugins/net_plugin/net_plugin.cpp

```diff
@@ -1221,12 +1340,12 @@ namespace eosio {
       buffer_queue.fill_out_buffer( bufs );
 
       strand.post( [c{std::move(c)}, bufs{std::move(bufs)}]() {
-         boost::asio::async_write( *c->socket, bufs,
-            boost::asio::bind_executor( c->strand, [c, socket=c->socket]( boost::system::error_code ec, std::size_t w ) {
+         //capture by value is const so we need mutable keyword
+         auto write_lambda = [c, socket=c->socket]( boost::system::error_code ec, std::size_t w ) mutable {
```

Comment:
Good point. I missed that this was copying the flex_socket by value. It would be more efficient to only have one shared_ptr copy instead of two.

Reply:
unfortunately we can't cast `tcp::socket` to `ssl::stream<tcp::socket>` that is why we have two pointers.
do you have any suggestions? perhaps we can use `variant<socket, ssl_socket>` instead of two pointers. anyway we instantiate only one of them.

Reply:
You could use a shared_ptr<flex_socket> and then just unique_ptr or variant inside flex_socket.

Reply:
since we care about increasing reference count in `flex_socket`'s object using `shared_ptr` for `flex_socket` looks redundant. I think using variant inside `flex_socket` will do the trick to avoid two shared_ptr copies (hence only one reference count increase which is costly atomic operation)